### ISSUE TITLE
using different shops in parallel. 

### ISSUE THREAD

**Issue by NafaaBout** [Shopify/shopify_api#259]
We have a shopify app that can be installed and used by multiple shops to push products from our database to the shops in question. We do this using sidekiq jobs. and because Sidekiq uses threads to run jobs in parallel. so activating a session to access a shop doesn't work well. because the session activation happen on the class `ShopifyAPI::Base` level. so when two shops are pushing at the same time. sometimes we get data mixed up. one shop gets the data of another one.
so if I do this
```
sess = ShopifyAPI::Session.new(shop.domain, shop.token)
ShopifyAPI::Base.activate_session(sess)
# and do some staff here
product = ShopifyAPI::Product.new(product_attributes)
product.save 
ShopifyAPI::Base.clear_session
```
So before the product saves to the `shop x` in thread 1, the thread 2 may activate the session for a different `shop y`, so the product in thread 1 will be saved to the `shop y` of the thread 2.
I've seen it happen randomly in our app although we are using MRI. I expect it to be worse on JRuby or Rubinius, where there is real threading.

**Comment by kevinhughes27** [comment]
There is also a threadsafe fork of activeresource you can add which can solve this.

also have you tried: 
```
ShopifyAPI::Session.temp(shopify_domain, token) do
end
```

**Comment by NafaaBout** [comment]
Oh, thanks. didn't know about `temp`.

**Comment by julionc** [comment]
If you are using ShopifyApp, you can take a look on [Shop.with_shopify_session ](https://github.com/Shopify/shopify_app/blob/master/lib/shopify_app/shop.rb#L10)method

**Comment by heldopslippers** [comment]
Hi Guys, 

We just discovered we have the same problem (using Shop.with_shopify_session in Sidekiq).
Sadly we discovered it while having a lot of installs right now. So this is gonna be a pain to fix :P 

Obviously it is hard to test these kind of things (I don't know how). 
But Will give: 
`ShopifyAPI::Session.temp(shpoify_domain, token)` a go, see if that fixes things. 

Does anybody have another solution?
